lib/workflow/task-graph.js

```diff
@@ -0,0 +1,493 @@
+            if (_.contains(value, Constants.TaskStates.Finished)) {
+                value = Constants.FinishedTaskStates;
+                task.waitingOn[dep] = [Constants.TaskStates.Finished];
+            }
+            // The dependent task is no longer terminal on state <value> since the
+            // current task lists that condition as a dependency.
+            // task.nonTerminalOnStates gets substracted from task.terminalOnStates
+            // when we traverse to the dependent task.
+            nonTerminalOnStates = Array.isArray(value) ? value : [value];
+            return self._visitGraphNode(dep, task.injectableName, markers, nonTerminalOnStates);
+        });
+
+        marker.permanentMark = true;
+        marker.temporaryMark = false;
+    };
+
+    TaskGraph.prototype.detectCyclesAndSetTerminalTasks = function() {
+        var self = this;
+        var markers = {};
+        _.forEach(self.tasks, function(task, taskId) {
+            self._visitGraphNode(taskId, null, markers, null);
+        });
+    };
+
+    /*
+     * Take the tasks definitions in this.definition.tasks, generate instanceIds
+     * to use for each task, and then create new Task objects that reference
+     * the instanceIds in their dependencies instead of user-created task labels.
+     */
+    // TODO: Replace this with a proper DFS traversal instead of iterating
+    TaskGraph.prototype._populateTaskData = function() {
+        var self = this;
+
+        assert.arrayOfObject(self.definition.tasks);
+        var idMap = _.transform(self.definition.tasks, function(result, v) {
+            result[v.label] = uuid.v4();
+        }, {});
+        return Promise.map(self.definition.tasks, function(taskData) {
+            assert.object(taskData);
+            if (_.has(taskData, 'taskName')) {
+                assert.string(taskData.taskName);
+            } else if (_.has(taskData, 'taskDefinition')) {
+                assert.object(taskData.taskDefinition);
+            } else {
+                throw new Error("All TaskGraph tasks should have either a taskName" +
+                    " or taskDefinition property.");
+            }
+
+            _.forEach(_.keys(taskData.waitOn), function(waitOnTask) {
+                var newWaitOnTaskKey = idMap[waitOnTask];
+                assert.ok(newWaitOnTaskKey, 'Task to wait on does not exist: ' + waitOnTask);
+                var waitOnTaskValue = taskData.waitOn[waitOnTask];
+                delete taskData.waitOn[waitOnTask];
+                taskData.waitOn[newWaitOnTaskKey] = waitOnTaskValue;
+            });
+            var taskOverrides = {
+                instanceId: idMap[taskData.label],
+                waitingOn: taskData.waitOn,
+                ignoreFailure: taskData.ignoreFailure
+            };
+
+            if (taskData.taskName) {
+                return self.constructTaskObject(taskData.taskName, taskOverrides,
+                        taskData.optionOverrides, taskData.label)
```

Comment:
As a personal style preference, I like to keep function arguments further indented away from the part of the call that doesn't include the function name, just because it's clearer to me when reading that they are arguments to the previous line, and not a new line of logic or anything. I acknowledge that it's personal preference and am curious what others prefer in these cases. I don't think we have any sort of standardized style rules within our codebase pertaining to this corner case, but do you think we would benefit from it?